zerver/openapi/zulip.yaml

```diff
@@ -857,15 +857,17 @@ paths:
       - name: emoji_name
         in: query
         description: |
-                     Name of the emoji you want to add as as a reaction.
+                     Name of the emoji you want to add as a reaction.
 
                      To find an emoji's name, hover over a message to
-                     reveal three icons on the right, then click the smiley face icon.
-                     Images of available reaction emojis appear. Hover over the emoji
-                     you want, and note that emoji's text name.
+                     reveal three icons on the right, then click the smiley face icon. Images of available reaction emojis appear.
```

Comment:
[Line 863] Please indent this line after the end of the first sentence. The line seems to stick out otherwise.

Reply:
Hi @Hypro999 

Hi Hemanth,

Thanks so much for your review. As you suggested, I made changes to the following files on my working branch on my remote dev instance:

zerver/openapi/zulip.yaml
templates/zerver/help/include/rest-endpoints.md
zerver/openapi/python_examples.py

I ran

tools/lint
tools/test-backend zerver.tests.test_openapi

and they both passed.

So now I need help on the right way to show those changes via a pull request -- I definitely need help on the git stuff!

I think I should do something like:

git add to these files:

zulip.yaml, rest-endpoints.md, and python_examples.py

Then git commit --amend to redo my commit msg.

Then what exactly for the second git pull request?

Do I just do something like:

$ git push origin issue-123

or do I have to do something else, seeing as I already have one pull request on this?

Thanks!

Laurel

Reply:
@laurel-personal you're definitely on the right track here :+1: . What I suggest doing (step-by-step) would be:
1. Go to the `zulip` folder you cloned.
2. run `git add -A`, this will add all of your changes to the "staging area".
3. run `git status` just to check that you haven't added or modified any extra files by accident.
4. run `git commit --amend`, this will modify the last commit you made to include these new changes. This also presents you with an opportunity to change your commit message. So update the commit message here (first line for summary and then second line onward you can add further explanation of what you did if you'd like).
5. Now you can (force) push the changes back to this same branch on your fork. To do this, you can run `git push --force origin add-emoji-to-msg-4`. That will automatically update this Pull Request (PR for short) as well. This pull request dynamically looks at what changes have been made on your `add-emoji-to-msg-4` branch and not just what changes were there when you first opened your pull request. The reason you will need to force push and not just normally push is that you're changing and existing commit and not adding a new one (which is what you want to do right now). This means that the git history on your local copy of `zulip` and remote (GitHub) copy of `zulip` have diverged down two different paths. The `--force` option will force the remote copy to overwrite it's current history with the one that you are pushing. This option can *sometimes* be dangerous, but that's only if one starts getting very careless with this option, so no need to worry about it.

This should be enough to cleanly publish your changes back to this PR for further review. :slightly_smiling_face:

Reply:
On a side-note, I strongly suggest taking a day or two to get really comfy with git. It's **really** useful and not just for programming. I use git for everything ranging from development/programming/other technical things to simple things such as keeping track of when I'm backing up which photos from my phone (for example). It can be be very useful in a lot of day-to-day things.

Here are some resources you can start off with:

Tutorials [text-based]:
1. [Atlassian's tutorials](https://www.atlassian.com/git/tutorials)
2. [A good medium article by George Seif](https://medium.com/@george.seif94/a-full-tutorial-on-how-to-use-github-88466bac7d42)

Video Tutorial:
1. [Corey Schafer's series](https://www.youtube.com/watch?v=HVsySz-h9r4&list=PL-osiE80TeTuRUfjRe54Eea17-YfnOOAx)

Documentation and Quick Reference:
1. [The Official Docs](https://git-scm.com/docs/gittutorial)
2. [Tutorialspoint](https://www.tutorialspoint.com/git/index.htm)

Reply:
Hi @Hypro999  
 
Once again, thanks so much for your reply -- I really appreciate your help!

So, before moving ahead with anything else, I seem to have introduced an error in `zerver/openapi/python_examples.py`. When I now run `tools/lint`, I get this error:
 
 ```
 zerver/openapi/python_examples.py:538: error: "Client" has no attribute "add_emoji_to_message"
 See https://zulip.readthedocs.io/en/latest/testing/mypy.html

 
 ```
 
 Here is what I added to `zerver/openapi/python_examples.py`
 
 ```
 
 @openapi_test_function("/messages/{message_id}/reactions:post")
 def add_emoji_to_message(client):
     # type: (Client) -> None
 
     # {code_example|start}
     # Add emoji to messeage
     request = {
         'emoji_name': 'working_on_it'
     }
     result = client.add_emoji_to_message(request)
     # {code_example|end}
 
     validate_against_openapi_schema(result, '/messages/{message_id}/reactions', 'post', '200')
 
 ```
 
 What am I doing wrong?
 
 Thanks.

Reply:
I think the issue here is that the `client` object is part of https://github.com/zulip/python-zulip-api, and it doesn't have an `add_emoji_to_message` function.  The function there is called `add_reaction`, and that's what you should call here.

Reply:
I think I did what Tim said (call `add_reaction`, but I must be doing something wrong.

Here are my new error messages from `tools/lint`

```
(zulip-py3-venv) ~/zulip (add-emoji-to-msg-4)$ tools/lint
mypy      | zerver/openapi/python_examples.py:630: error: Name 'add_reaction' already defined on line 529
mypy      |
mypy      | See https://zulip.readthedocs.io/en/latest/testing/mypy.html for debugging tips.
pyflakes  | zerver/openapi/python_examples.py:630: redefinition of unused 'add_reaction' from line 529
(zulip-py3-venv) ~/zulip (add-emoji-to-msg-4)$

```


Starting at line 529 -- this is what I added, based on Tim's suggestion to call `add_reaction`

```

@openapi_test_function("/messages/{message_id}/reactions:post")
def add_reaction(client):
    # type: (Client) -> None

    # {code_example|start}
    # Add emoji to messeage
    request = {
        'emoji_name': 'working_on_it'
    }
    result = client.add_reaction(request)
    # {code_example|end}

    validate_against_openapi_schema(result, '/messages/{message_id}/reactions', 'post', '200')
    
```

***
Here is what starts at line 630

```

def add_reaction(client, message_id):
    # type: (Client, int) -> None
    request = {
        'message_id': str(message_id),
        'emoji_name': 'joy',
        'emoji_code': '1f602',
        'emoji_type': 'unicode_emoji'
    }
    result = client.add_reaction(request)

    assert result['result'] == 'success'
    
```

What am I doing wrong?

Reply:
@laurel-personal, the direct cause of this error is the fact that a function called `add_reaction` already exists in that file (`zerver/openapi/python_examples.py`) and you are redefining it with the code you added starting from line 529.

Now as for the right course of action from here, and just a warning I'm not 100% sure of this, but I think that it might be to replace what was previously there with your code since your code uses the new and recommended `validate_against_openapi_schema` method like most of the other functions.

Maybe @timabbott could quickly verify if this is indeed the right thing to do.

Reply:
That plan sounds right to me.